src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/ObjectHeader.cs

```diff
@@ -61,30 +83,37 @@ public static unsafe int GetHashCode(object o)
             fixed (MethodTable** ppMethodTable = &o.GetMethodTableRef())
             {
                 int* pHeader = GetHeaderPtr(ppMethodTable);
-                int bits = *pHeader;
-                int hashOrIndex = bits & MASK_HASHCODE_INDEX;
-                if ((bits & BIT_SBLK_IS_HASHCODE) != 0)
-                {
-                    // Found the hash code in the header
-                    Debug.Assert(hashOrIndex != 0);
-                    return hashOrIndex;
-                }
-
-                if ((bits & BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX) != 0)
+                int hashCode = TryGetHashCode(pHeader);
+                if (hashCode != 0)
                 {
-                    // Look up the hash code in the SyncTable
-                    int hashCode = SyncTable.GetHashCode(hashOrIndex);
-                    if (hashCode != 0)
-                    {
-                        return hashCode;
-                    }
+                    return hashCode;
                 }
 
                 // The hash code has not yet been set.  Assign some value.
                 return AssignHashCode(o, pHeader);
             }
         }
 
+        /// <summary>
+        /// If a hash code has been assigned to the object, it is returned. Otherwise zero is
+        /// returned.
+        /// </summary>
+        /// <remarks>
+        /// This method needs to follow the rules in RestrictedCallouts.h, as it may be called
```

Comment:
I do not think compliancy with `RestrictedCallouts` is an explicit contract or need to be mentioned in doc comments. 
It is a nice and mostly unavoidable implementation detail, but the API is meaningful on its own.

Reply:
I removed the comments about following the restricted call rules and instead fixed the test that exercises this functionality. Though it looks like these tests are not being run for NativeAOT yet. #67743

I added a NativeAOT smoke test that tests this scenario. Since GC restricted callouts are not a public API, I had to copy in some bits of the NativeAOT CoreLib into the test. Is there a better way to test this?

Reply:
We try to keep the number of the nativeaot smoke test projects small. Native aot smoke tests run by default in the CI and each project takes 10's of seconds to compile and run in CI, so the time to run can add up quickly. The smoke tests are not meant to provide full coverage.

I think it would be better to beef up `src/tests/Interop/ObjectiveC/ObjectiveCMarshalAPI/Program.cs` to catch this sort of problems: allocate a pile of objects that are eligible for tracking, create tracking handle for some of them, take orphaned lock on some of them, and I assume that we are going to start running this test regularly in the CI soon.

Reply:
I reverted the extra smoke test. The `ObjectiveCMarshalAPI` test now creates objects with the object header in various different states.